refactor: rewrite to esm

Julien-R44 opened the pull request:
So here is my attempt to a rewrite for v6. If you have the time and inclination I'd be more than happy with a review in which you are very "picky", so that for the next PRs I'll know exactly what to do to match your style (which has evolved quite a bit since V6  )

---

I removed the tags for the asset manager as planned. They will go in their own package @adonisjs/vite and @adonisjs/webpack respectively.

What is missing : 

- REPL Bindings. ( @adonisjs/repl needs to be rewrited )
- Globals for @adonisjs/drive ( @adonisjs/drive needs to be rewrited )
- Global `env`. I'm not sure how to approach this since there is no way to access the env at runtime from a third-party package without full parsing it. Maybe we can do that (full parsing of the .env file in the provider, but it doesn't seem like a great solution). If not, maybe we can get rid of this global ? it doesn't seem very important to me.

--- 

I've exposed a "service", as you did in the core, which allows to get the View from the container when ready :

```ts
import { view } from '@adonisjs/view/services/view'
view.render(...)
```

It seemed a bit verbose so I re-exported that from the root as well :
```ts
import { view } from '@adonisjs/view'
view.render(...)
```

The provider is accessible via the following import and must be added as is in the .adonisrc.json `@adonisjs/view/providers/view_provider.js`

---

We will need to update @adonisjs/http-server to use the latest version of `macroable`, and then update @adonisjs/core
The PR is made on the "next" branch and can therefore be merged without issues

And I think that's pretty much all! The framework looks really super modern with ESM and this new container system. A lot of changes, I had a little trouble at first, but I'm getting the hang of it now.

As I was saying: don't hesitate to be very "nitpicky" on your review, whether it's about a file in the wrong place, a bad casing or anything else. I'll be happy to fix it and remember it for the next ones !

Julien-R44 committed:
refactor: rewrite to esm

Julien-R44 commented on providers/view_provider.ts:
```diff
@@ -0,0 +1,152 @@
+   */
+  private registerHTTPContextGetter(view: ViewContract) {
+    // Check if getter is already defined
+    if (Object.getOwnPropertyDescriptor(HttpContext.prototype, 'view')) {
```
This check could be removed if we put `configurable: true` in the Object.defineProperty of `macroable` as explained on Discord

Julien-R44 commented on test/view_provider.spec.ts:
```diff
@@ -0,0 +1,90 @@
+  test('todo: register env global')
+  test('todo: register repl binding')
+  test('todo: do not register repl binding when not in repl environment')
+  test('todo: register driveUrl and driveSignedUrl globals')
```
I left some tests in todo for the features I haven't added yet

Julien-R44 commented on test_helpers/index.ts:
```diff
@@ -0,0 +1,42 @@
+    .withCoreConfig()
+    .withCoreProviders()
+    .merge({ config: additionalConfig })
+    .create(BASE_URL, { importer: IMPORTER })
```
It seems to me that it is, but I'm curious to know if this is now the desired approach to create and test an application

thetutlage replied:
Yup, this is the way to go. Each package exposes factories that can be used for testing. Lemme know if you think any part of these factories can be improved

Julien-R44 replied:
Yes this is super useful !

I would say that the only thing that seems a bit cumbersome is to add an `importer`. Maybe we could have a default one ?

thetutlage replied:
The default one shipped via package won't work for relative imports. In case of relative imports, if the `import` method is defined inside a package, then it will look for files relative to the package and not the host app.

Alternative is to use `import.meta.resolve` API. But that is experimental right now and also a little slow on performance

Julien-R44 commented:
I just remembered that I forgot to export a `defineConfig` for the package config. Which only contains `cache: { enabled }` for now

Also to use `#` to define private methods.

I'll fix this asap

thetutlage commented on package.json:
```diff
@@ -1,27 +1,41 @@
 {
 	"name": "@adonisjs/view",
-	"version": "6.2.0",
+	"version": "7.0.0-0",
```
This should be bumped when we create a release using `np`. Basically, `np` changes the version number, we never do it manually

thetutlage commented on package.json:
```diff
@@ -1,27 +1,41 @@
 		"version": "npm run build",
-		"sync-labels": "github-label-sync --labels ./node_modules/@adonisjs/mrm-preset/gh-labels.json adonisjs/view",
+		"sync-labels": "github-label-sync --labels .github/labels.json adonisjs/logger",
+		"vscode:test": "node --loader=ts-node/esm bin/test.ts",
```
I am starting to call it `quick:test`. The `vscode:test` was a weird name that I choose because I was not able to decide on the name.

But, `quick:test` is something we will have in the starter kit of official packages

thetutlage commented on package.json:
```diff
@@ -41,62 +55,37 @@
-	"np": {
-		"contents": ".",
-		"anyBranch": false
+		"@adonisjs/core": "^6.0.0"
```
This might fail the semver check. So we will have to use `^6.1.5-2`. I use this app to see which expression matches which versions. https://semver.npmjs.com/

thetutlage commented on providers/view_provider.ts:
```diff
@@ -0,0 +1,152 @@
+  /**
+   * Add globals for resolving routes
+   */
+  private addRouteGlobal(view: ViewContract, router: HttpRouterService) {
```
Not using the `private` modifier anymore. Instead using `#` for creating JavaScript private properties

thetutlage commented on providers/view_provider.ts:
```diff
@@ -0,0 +1,152 @@
+   */
+  private async shouldCacheViews(): Promise<boolean> {
+    const config = await this.app.container.make('config')
+    const cacheViews = config.get('views.cache', true)
+
+    return typeof cacheViews === 'string' ? cacheViews === 'true' : (cacheViews as boolean)
```
If we are pulling the `cache` value from config. We can be sure that it will be a boolean always

thetutlage commented:
Everything looks good to me. There might be couple of places that needs cleanup that are hard to spot in the PR diff view. So, please bear with me if I open a new PR with those changes and explain them to you.

---

Regarding: "It seemed a bit verbose so I re-exported that from the root as well :"

I will advise not to do that. Because, services are not pure. Meaning, a service import will break if the app is not ready.

I want the package exports to be pure. Meaning, they work as expected, regardless of where you import them. Services behave different if you export them without booting the app and after booting the app. So, I want this behavior to be limited to the `services` files only.

This way, we will be able to explain properly in docs about the behavior of services and tell everyone that as a rule of thumb, you should expect services to be broken, unless you have a booted application in place.

I hope this make sense

Julien-R44 committed:
refactor: revert bump version

Julien-R44 committed:
refactor: vscode:test to quick:test

Julien-R44 committed:
refactor: update semver peerdeps adonisjs/core

Julien-R44 committed:
refactor: use `#` for private fields

Julien-R44 committed:
refactor: remove type checks on cacheViews

Julien-R44 committed:
chore: remove check for already register getter

Julien-R44 committed:
refactor: dont export view service from root

Julien-R44 committed:
chore: update adonisjs/core

Julien-R44 committed:
feat: add defineConfig helper

Julien-R44 commented:
Awesome, thanks for the review. I have fixed everything. Don't worry if you want to change some more things, I'll be happy to look at your changes 💪 

You are right about re-exporting the service from the root. Thanks for the clarification, I deleted the export

Also added `defineConfig` helper

Julien-R44 committed:
chore: add header

thetutlage commented:
Cool. Looks good 👍
